cloudcopy/server/models/workflow.py

```diff
@@ -0,0 +1,178 @@
+            'null': True,
+            # UUID value of task ID
+            # representing the next scheduled invocation of this workflow
+        },
+        'running_jobs': {
+            'type': 'integer',
+            'default': 0
+        },
+        'concurrency': {
+            'type': 'integer',
+            'null': True,
+            'default': 0
+            # number of simultaneous job executions
+            # default of 0 = unlimited
+            # if a workflow is triggered when the concurrency limit is reached,
+            # the job will immediately fail
+        },
+        'cooldown': {
+            'type': 'integer',
+            'null': True,
+            'default': 0
+            # time (seconds) to wait before the next execution
+            # after the previous one has finished
+        },
+        'created': {
+            'type': 'text',
+            'created': True
+        },
+        'updated': {
+            'type': 'text',
+            'updated': True
+        }
+    }
+
+    async def resolve_steps(self, steps):
+        """Get workflow.steps as an object
+
+        Resolve all database references, either by ID or name
+        """
+        if not steps:
+            return None
+
+        steps = json.loads(steps) if isinstance(steps, str) else steps
+        resolved = {}
+        db = await Database.initialize(self._database)
+        for step in steps:
+            # look for database references and resolve them
+            for key in ('source', 'target'):
+                if key not in step:
+                    continue
+                value = step[key]
+                if is_url(value):
+                    continue
+                if value in resolved:
+                    step[key] = resolved[value]
+                else:
+                    # memoize lookups
+                    try:
+                        resolved[value] = step[key] = await db.get_field(
+                            'url', value
+                        )
+                    except Exception as e:
+                        raise ValueError(
+                            f'Error resolving {key} for step: {step}\nMessage: {e}'
+                        )
+        return steps
+
+
+    async def pre_add(self, query):
+        # validate steps
+        parent = super()
+        if hasattr(parent, 'pre_add'):
+            # perform base manipulations/checks
+            await parent.pre_add(query)
+
+        # perform validation of "steps"
+        values = query.data('values')
+        if isinstance(values, list):
+            return await self.pre_add_many(values)
+        else:
+            return await self.pre_add_one(values)
+
+    async def pre_add_many(self, values):
+        for value in values:
+            await self.pre_add_one(value)
+
+    async def pre_add_one(self, values):
+        steps = values.get('steps')
+        await self.resolve_steps(steps)
+
+    async def post_add_record(self, record, result):
```

Comment:
Main integration point between API and tasks demonstrated here:
1) in API post-action handlers (`post_add_record`), execute or schedule the task (depending on whether it is periodic)
2) in tasks, define an executable task that represents async follow-up work